earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +151,137 @@ class khMTSet
     khLockGuard guard(mutex);
     return std::vector<T>(set.begin(), set.end());
   }
+  void clear(void) {
+    khLockGuard guard(mutex);
+    set.clear();
+  }
+  size_t size(void) {
+    return set.size();
+  }
+};
+
+
+template <class K, class V>
+class khMTMap
+{
+ protected:
+  typedef std::unordered_map<K,V> Map;
+  Map     map;
+  mutable std::mutex mutex;
+
+ private:
+  // private and unimplimented
+  khMTMap& operator=(const khMTMap&) = delete;
+
+  khMTMap(const khMTMap& o, const std::lock_guard<std::mutex>&) : map(o.map) { }
+  khMTMap(khMTMap&& o, const std::lock_guard<std::mutex>&) : map(std::move(o.map)) { }
+ public:
+  khMTMap(void) { }
+  khMTMap(const khMTMap& o) : khMTMap(o, std::lock_guard<std::mutex>(o.mutex)) { }
+  khMTMap(khMTMap&& o) : khMTMap(o, std::lock_guard<std::mutex>(o.mutex)){ }
+
+  void insert(const std::pair<K,V> &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.insert(val);
+  }
+  void erase(const K &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.erase(val);
+  }
+  size_t size(void) {
+    return map.size();
+  }
+  V& operator[](const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map[key];
+  }
+  void clear(void) {
+    std::lock_guard<std::mutex> guard(mutex);
+    map.clear();
+  }
+  std::vector<K> keys(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    std::vector<K> thekeys;
+    thekeys.reserve(map.size());
+    for (const auto& p : map) {
+      thekeys.push_back(p.first);
+    }
+    return thekeys;
+  }
+  bool contains(const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.find(key) != map.end();
+  }
+  bool empty(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.empty();
+  }
 };
 
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but inheriting from vector publicly isn't great
+template <typename T>
+class MTVector : public std::vector<T> {
+  private:
+    mutable std::mutex mtx;
+
+    MTVector(const MTVector& a, const std::lock_guard<std::mutex> &) : Base(a) { }
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
+    void push_back(const T& v) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::push_back(v);
+    }
+    MTVector() : Base() { }
+    // forward to private copy constructor to protect vector from modification
+    MTVector(const MTVector& a) : MTVector(a, std::lock_guard<std::mutex>(a.mtx)) { }
+
+    const T& operator[] (const size_t idx) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return Base::operator[](idx); 
+    }
+
+    bool operator==(const MTVector& x) const {
+      std::mutex* firstMtx = &x.mtx;
```

Comment:
Good work here - I'm not sure I would have thought of locking both mutexes and acquiring them in a fixed order to avoid deadlock. Might be good to include a comment that that's what you're doing.